### PR TITLE
URL priorisation, split meta WARCs, and miscellaneous bug fixes

### LINKED ISSUE #332: ImportError with html5lib 0.99999999 (2016-07-14)
html5lib 0.99999999 renamed `html5lib.tokenizer`:

> Move a whole load of stuff (inputstream, ihatexml, trie, tokenizer, utils) to be underscore prefixed to clarify their status as private.

https://pypi.python.org/pypi/html5lib

and thus wpull crashes with ImportError on startup:

```
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/libgrabsite/main.py", line 368, in main
    wpull.__main__.main()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/wpull/__main__.py", line 29, in main
    builder.build()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/wpull/builder.py", line 195, in build
    self._build_html_parser()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/wpull/builder.py", line 692, in _build_html_parser
    from wpull.document.htmlparse.html5lib_ import HTMLParser
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/wpull/document/htmlparse/html5lib_.py", line 3, in <module>
    import html5lib.tokenizer
ImportError: No module named 'html5lib.tokenizer'
```

(That's a traceback from wpull 1.2.3, but it looks like wpull 2.0.1 has the same import.)

### LINKED ISSUE #340: Unable to parse URL with empty port number
From a recent log: wpull.url - WARNING - Unable to parse URL ‘https://www.themelvins.net:’: invalid literal for int() with base 10: ''.  Note the colon at the end of the URL.  I would expect it would use the default port number for the protocol in this case.

### LINKED ISSUE #355: sitemap parsing/discovery seems broken for www.zeit.de
`wpull --recursive --sitemaps www.zeit.de` leads to gazillion of errors like

```
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=5881’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=5881'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=12847’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=12847'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=3160’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=3160'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=4196’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=4196'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=8086’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=8086'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=370’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=370'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=611’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=611'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=10841’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=10841'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=12553’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=12553'.
WARNING Unable to parse URL ‘http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=86’: URL contains control codes: 'http://www.zeit.de/gsitemaps/\n\t\t\t\t\thttp://www.zeit.de/gsitemaps/index.xml?page=86'.
```

As a workaround I tried to reject those URLS with "--reject-regex 'gsitemaps\/index.xml\?page='" and "--exclude-directories gsitemaps" but neither worked. I am not sure if it is parsing them or trying/accessing them so that might make sense.

```
$ python3 --version
Python 3.4.2
$ ~/.local/bin/wpull3 --version
psutil: No module named 'psutil'. Resource monitoring will be unavailable.
2.0.1

```

### LINKED ISSUE #376: TypeError: Tuple or struct_time argument required
*What I wanted:* Webpage downloaded
 
*What I expect:* Webpage downloaded

*What happened:* Wpull crashed

*The command or website causes the problem:* https://www.wired.com/geekdad/tag/podcast/

*Operating system:* Archlinux, Debian

*Python version:* 3.6.2, 3.4.2

*Wpull version:* 2.0.1

*Log/Output:*

```
INFO Fetching ‘https://www.wired.com/geekdad/tag/podcast/’.
INFO Fetched ‘https://www.wired.com/geekdad/tag/podcast/’: 200 OK. Length: 19698 [text/xml; charset=UTF-8].
ERROR Fatal exception.
Traceback (most recent call last):
  File "/home/me/.local/lib/python3.4/site-packages/wpull/application/app.py", line 152, in run
    yield from pipeline.process()
  File "/home/me/.local/lib/python3.4/site-packages/wpull/pipeline/pipeline.py", line 194, in process
    yield from self._process_one_worker()
  File "/home/me/.local/lib/python3.4/site-packages/wpull/pipeline/pipeline.py", line 215, in _process_one_worker
    task.result()
  File "/usr/lib/python3.4/asyncio/futures.py", line 277, in result
    raise self._exception
  File "/usr/lib/python3.4/asyncio/tasks.py", line 235, in _step
    result = coro.send(value)
  File "/home/me/.local/lib/python3.4/site-packages/wpull/pipeline/pipeline.py", line 119, in process
    item = yield from self.process_one(_worker_id=worker_id)
  File "/home/me/.local/lib/python3.4/site-packages/wpull/pipeline/pipeline.py", line 103, in process_one
    yield from task.process(item)
  File "/home/me/.local/lib/python3.4/site-packages/wpull/application/tasks/download.py", line 492, in process
    yield from session.app_session.factory['Processor'].process(session)
  File "/home/me/.local/lib/python3.4/site-packages/wpull/processor/delegate.py", line 29, in process
    return (yield from processor.process(item_session))
  File "/home/me/.local/lib/python3.4/site-packages/wpull/processor/web.py", line 92, in process
    return (yield from session.process())
  File "/home/me/.local/lib/python3.4/site-packages/wpull/processor/web.py", line 186, in process
    yield from self._process_loop()
  File "/home/me/.local/lib/python3.4/site-packages/wpull/processor/web.py", line 245, in _process_loop
    exit_early, wait_time = yield from self._fetch_one(cast(Request, self._item_session.request))
  File "/home/me/.local/lib/python3.4/site-packages/wpull/processor/web.py", line 309, in _fetch_one
    action = self._handle_response(request, response)
  File "/home/me/.local/lib/python3.4/site-packages/wpull/processor/web.py", line 422, in _handle_response
    filename = self._file_writer_session.save_document(response)
  File "/home/me/.local/lib/python3.4/site-packages/wpull/writer.py", line 306, in save_document
    self.set_timestamp(self._filename, cast(HTTPResponse, response))
  File "/home/me/.local/lib/python3.4/site-packages/wpull/writer.py", line 144, in set_timestamp
    last_modified = time.mktime(last_modified)
TypeError: Tuple or struct_time argument required
CRITICAL Sorry, Wpull unexpectedly crashed.
CRITICAL Please report this problem to the authors at Wpull's issue tracker so it may be fixed. If you know how to program, maybe help us fix it? Thank you for helping us help you help us all.
INFO Exiting with status 1.

```

### LINKED ISSUE #377: Fails to properly parse URLs with the wrong slash
*What I wanted:* To download pages from Naval Vessel Register
 
*What I expect:* To have properly downloaded pages from the Navel Vessel Register

*What happened:* Pages returned 404

*The command or website causes the problem:* wpull http://www.nvr.navy.mil/QUICKFIND/SHIPSDETAIL_ACTIVE_H.HTML --recursive --level 1 --no-check-certificate --no-robots --ignore-fatal-errors --no-directories --wait 0.5 --random-wait --content-on-error --output-file output.txt

*Operating system:* Windows 10

*Python version:* Python 3.5.2

*Wpull version:* 2.0.1

*Log/Output:* 
```
2017-09-24 01:17:43,254 - wpull.processor.web - INFO - Fetching ‘http://www.nvr.navy.mil/QUICKFIND/..\SHIPDETAILS\SHIPSDETAIL_DDG_97_1770.HTML’.
2017-09-24 01:17:43,318 - wpull.processor.web - INFO - Fetched ‘http://www.nvr.navy.mil/QUICKFIND/..\SHIPDETAILS\SHIPSDETAIL_DDG_97_1770.HTML’: 404 Not Found. Length: 3800 [text/html].
2017-09-24 01:17:43,836 - wpull.processor.web - INFO - Fetching ‘http://www.nvr.navy.mil/QUICKFIND/..\SHIPDETAILS\SHIPSDETAIL_SSN_776_1977.HTML’.
2017-09-24 01:17:43,900 - wpull.processor.web - INFO - Fetched ‘http://www.nvr.navy.mil/QUICKFIND/..\SHIPDETAILS\SHIPSDETAIL_SSN_776_1977.HTML’: 404 Not Found. Length: 3800 [text/html].
2017-09-24 01:17:44,513 - wpull.processor.web - INFO - Fetching ‘http://www.nvr.navy.mil/QUICKFIND/..\email.HTML’.
2017-09-24 01:17:44,598 - wpull.processor.web - INFO - Fetched ‘http://www.nvr.navy.mil/QUICKFIND/..\email.HTML’: 404 Not Found. Length: 3800 [text/html].
```

wpull seems to interpret URLs with the wrong slash differently than browsers. ArchiveBot job 1xruq9x9x3kkcmahie8vdrrbz is a case where this posed a serious problem.

### LINKED ISSUE #383: Plugins activated too late
While [wpull 1 loaded plugins almost immediately](https://github.com/chfoo/wpull/blob/v1.2.3/wpull/builder.py#L189-L190), they are only loaded/activated as [one of the last steps in the setup pipeline in version 2](https://github.com/chfoo/wpull/blob/d0a9dffd2f79e62242e5ddb603c5e77e5b012e91/wpull/application/builder.py#L178). As a result, it is not possible to cleanly extend or modify the functionality of many core components anymore.
For example, the plugin in ArchiveBot overrides the pragmas to disable syncing the SQLite database to disk. It does this by [subclassing `SQLiteURLTable`](https://github.com/ArchiveTeam/ArchiveBot/blob/23974b22ce475440309c7433161126f0b8ae2024/pipeline/archivebot/wpull/plugin.py#L18-L22) and [replacing the `URLTableImplementation` class by this subclass](https://github.com/ArchiveTeam/ArchiveBot/blob/23974b22ce475440309c7433161126f0b8ae2024/pipeline/archivebot/wpull/plugin.py#L71). However, this does not actually work: by the time the plugin is activated, the URL table instance already exists, so these changes don't have any effect.

The `PluginSetupTask` should be moved to the beginning, as was the case in wpull 1, to maximise customisability – ideally directly after the `LoggingSetupTask`.
It seems that the download_progress plugin will also need some changes as the `event_dispatcher` of `HTTPClient` and `FTPClient` is only available after instantiation. This could be solved by either subclassing the `Client` classes or by modifying the event dispatcher to allow adding listeners before instantiation.

### PR DESCRIPTION
**Pull request by JustAnotherArchivist**
This PR contains the various things I've implemented and fixed back in January and February on my fork, listed below.

Sorry for the huge PR. I was working based on version 2.0.3 (then only available on FalconK's fork) back then, and separating this into individual PRs for each implemented feature/fixed bug now would be really painful. Let me know if you want me to do it anyway.

* Add URL priorisation: `--priority-regex`/`--priority-domain` and the `get_priority` hook can be used to determine in which order URLs are retrieved.
* Load plugins early again, like in wpull 1.x. Before, plugins were loaded after everything was already initialised, which made it difficult to replace functionality by replacing the factory class map. (Fixes #383)
* Rewrite parts of the pipeline code to fix the order of URLs; this wasn't really a problem before, but it messed with the priorisation code.
* Fix `--concurrent` having no effect (#339)
* Pin html5lib version (fixes #332)
* Add splitting meta WARCs when the data WARC is split using `--warc-split-meta`
* Fix flattening of consecutive slashes in URLs (#380)
* Handle backslashes in URLs (fixes #377)
* Handle invalid dates in last modified date (fixes #376)
* Strip tab and newline characters from scraped URLs (fixes #355)
* Handle empty port correctly (fixes #340)

### CODE REVIEW

**Comment by ivan** [comment]
You might have noticed already, but your tests are failing in at least

```
  File "/home/travis/build/ArchiveTeam/wpull/thematrix/wpull/testing/integration/priorisation_test.py", line 161, in test_app_priority_plugin_get_urls_with_priorities
    self.assertEqual(builder.factory['Statistics'].files, 11)
nose.proxy.AssertionError: 5 != 11
```

Some other tests are probably failing from the earlier merge. Because it would be very tricky to continue test-driven development with failing tests, they should probably all be fixed (or some skipped?) as part of this PR (or in PR to land before this one if you wish.)

**Comment by JustAnotherArchivist** [comment]
I completely agree.

I remember some tests failing back in February, but definitely not that many. And yeah, some of the failures definitely come from @falconkirtaran's changes, e.g. the `unhashable type: dict` ones.

I'll look into it. To keep things separate, I will probably fix only the tests broken by my changes in this PR, then prepare a separate PR for the other broken tests.

**Comment by JustAnotherArchivist** [comment]
My memory was correct: I did get some test failures back in February, but not all of these. As it turns out, this is caused by a version difference in Tornado: my development venv had Tornado 4.5.1 whereas the Travis CI test used 4.5.3.

The problematic commit is <https://github.com/tornadoweb/tornado/commit/84bb2e285e15415bb86cbdf2326b19f0debb80fd> (thanks for finding this, @ivan). It replaces `localhost` with `127.0.0.1` in the return value of `tornado.testing.AsyncHTTPTestCase.get_url`, reasoning that due to "apparently a recent macos change, connecting to `localhost` when the ipv6 port is unbound now incurs a 200ms delay, slowing a full test run down by a factor of 20".

Because wpull treats `localhost` as a different host than `127.0.0.1`, all kinds of things in the test suite break (e.g. links to `localhost` get skipped, which causes the `AssertionError`s that the number of files doesn't match the expected value).

I'll override `tornado.testing.AsyncHTTPTestCase.get_url` in our subclass to always return a `localhost` URL. Perhaps we should switch to `127.0.0.1` in the future if that delay mentioned in Tornado's commit message becomes a problem.

**Comment by JustAnotherArchivist** [comment]
That looks much better. The remaining five errors are due to FalconK's changes.

@ivan, could you review please when you get a chance?

File: wpull/urlprioritiser.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        '''
+        return None
+
+    def _get_priority_from_hook(self, url_info: URLInfo, url_record: URLRecord) -> Union[int, None]:
+        '''Tries to call the get_priority hook. The hook can either return an int or None. If the hook is not connected, None is retured.'''
+        try:
+            # Only pass copies of the URLInfo and URLRecord instances to the hook to prevent modification.
+            # A deep copy is not actually needed as of writing this code because all attributes of both classes are immutable.
+            # However, to prevent future breakage if something changes in those classes, a deep copy is used anyway.
+            priority = self.hook_dispatcher.call(PluginFunctions.get_priority, copy.deepcopy(url_info), copy.deepcopy(url_record))
```
Comment (ivan) [inline wpull/urlprioritiser.py]: deepcopy is almost always a bad idea because it defers copying responsibility to a slowish and general algorithm that may copy more than needed, or in some rare cases less than needed.   Something that needs to be copied should usually implement e.g. its own `copy` method that copies just what is necessary.  Since you say `deepcopy` isn't needed right now, maybe just don't do it?

File: wpull/urlprioritiser_test.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+from wpull.application.plugin import PluginFunctions
+from wpull.pipeline.item import URLRecord
+from wpull.url import URLInfo
+from wpull.urlprioritiser import URLPrioritiser
+from wpull.urlfilter import RegexFilter
+
+
+PRIORITIES = [
+    (RegexFilter(accepted = r'//example\.com/'), 3),
```
Comment (ivan) [inline wpull/urlprioritiser_test.py]: no spaces around the `=`

File: wpull/warc/recorder.py [hunk 3]
```diff
@@ -112,13 +117,16 @@ def _check_journals_and_maybe_raise(self):
 
     def _start_new_warc_file(self, meta=False):
         '''Create and set as current WARC file.'''
-        if self._params.max_size and not meta and self._params.appending:
+        if self._params.max_size and (not meta or self._params.split_meta) and self._params.appending:
             while True:
-                self._warc_filename = self._generate_warc_filename()
+                self._warc_filename = self._generate_warc_filename(meta = meta)
```
Comment (ivan) [inline wpull/warc/recorder.py]: no spaces around the `=`

File: wpull/warc/recorder.py [hunk 4]
```diff
@@ -181,34 +192,81 @@ def _populate_warcinfo(self, extra_fields=None):
             bytes(info_fields) + b'\r\n')
         self._warcinfo_record.compute_checksum()
 
-    def _setup_log(self):
-        '''Set up the logging file.'''
+    def _flush_log(self, closing = False):
```
Comment (ivan) [inline wpull/warc/recorder.py]: no spaces around the `=`

File: wpull/warc/recorder.py [hunk 5]
```diff
@@ -181,34 +192,81 @@ def _populate_warcinfo(self, extra_fields=None):
...
+        '''Flush the logging
+
+        If there is already a logger and we want to split the meta WARC: write a meta WARC, delete the temporary log, and move the meta WARC if necessary.
```
Comment (ivan) [inline wpull/warc/recorder.py]: these lines are too long

File: wpull/warc/recorder.py [hunk 6]
```diff
@@ -181,34 +192,81 @@ def _populate_warcinfo(self, extra_fields=None):
...
+
+            log_record = WARCRecord()
+            log_record.block_file = gzip.GzipFile(
+                filename=self._log_temp_file.name
             )
-        )
+            log_record.set_common_fields('resource', 'text/plain')
+
+            log_record.fields['WARC-Target-URI'] = \
+                'urn:X-wpull:log'
```
Comment (ivan) [inline wpull/warc/recorder.py]: unnecessary line split?

File: wpull/warc/recorder.py [hunk 7]
```diff
@@ -364,45 +425,11 @@ def write_record(self, record):
...
-
-            self._log_temp_file = None
-
-            self._log_handler.close()
-            self._log_handler = None
-
-            if self._params.move_to is not None:
-                self._move_file_to_dest_dir(self._warc_filename)
+        if self._params.log:
+            self._flush_log(closing = True)
```
Comment (ivan) [inline wpull/warc/recorder.py]: no spaces around the `=`

File: wpull/warc/recorder_test.py [hunk 8]
```diff
@@ -327,6 +327,138 @@ def test_warc_recorder_max_size(self):
...
+    def test_warc_recorder_max_size_split_meta(self):
+        file_prefix = 'asdf'
+        cdx_filename = 'asdf.cdx'
+
+        warc_recorder = WARCRecorder(
+            file_prefix,
+            params=WARCRecorderParams(
+                compress=False,
+                extra_fields=[('Extra-field', 'my_extra_field')],
+                cdx=True, max_size=1,
```
Comment (ivan) [inline wpull/warc/recorder_test.py]: nitpicky but maybe put max_size on its own line as well

File: wpull/warc/recorder_test.py [hunk 9]
```diff
@@ -327,6 +327,138 @@ def test_warc_recorder_max_size(self):
...
+                split_meta = True,
+            )
+        )
+
+        # This produces three data and meta WARCs (00000 to 00002). The reason being that a new WARC is started after the second request.
```
Comment (ivan) [inline wpull/warc/recorder_test.py]: these lines are too long

File: wpull/warc/recorder_test.py [hunk 10]
```diff
@@ -327,6 +327,138 @@ def test_warc_recorder_max_size(self):
...
+
+        session = warc_recorder.new_http_recorder_session()
+        session.begin_request(request)
+        session.request_data(request.to_bytes())
+        session.end_request(request)
+        session.begin_response(response)
+        session.response_data(response.to_bytes())
+        session.response_data(response.body.content())
+        session.end_response(response)
+        _logger.info('Fetched http://example.com/1') # Note, the normal "Fetched" log message is also emitted before the session is closed. Otherwise, it'd end up in the next meta WARC.
```
Comment (ivan) [inline wpull/warc/recorder_test.py]: this line is too long, maybe just put the comment above it

**Comment by JustAnotherArchivist** [comment]
I just realised that my prioritisation code won't work as expected probably. It does assign priorities correctly, but it will still first process all todo URLs and only then go to errors. So if a prioritised URL errors out, it will still only be retried *after* all the other todos have been processed.

It should probably go after all highest priority URLs first, doing first the todos and then retrying the errors, and only then go to the next lower priority. Implementing this will probaby require some changes in how the pipeline and the database interact though. Currently, the pipeline requests an item with either status todo or error, i.e. the pipeline manages when to switch to retries, so to speak. Now, the database could return `NotFound` if the highest priority of errors is higher than that of todos, such that the pipeline switches to errors and processes those first. But that's a really awful hack.
Instead, the todo vs. error logic should probably be moved to the database, and the pipeline just retrieves an arbitrary item and schedules it for processing.

This is mostly a note for myself so I remember to look into this.

File: wpull/pipeline/pipeline.py [hunk 11]
```diff
@@ -75,31 +108,54 @@ def item_done(self):
...
 
+    @asyncio.coroutine
+    def drain(self):
+        '''Drain the queue: remove all items from the queue, consider them as completed, and cancel the producer futures.'''
+        while self._queue.qsize() > 0:
+            priority, entry_count, item_coro, producer_future = yield from self._queue.get()
+            yield from self.item_done()
+            if producer_future is not None:
+                producer_future.cancel()
+            _logger.debug('Drained {!r} ({!r}), {} remaining, {} unfinished'.format(item_coro, producer_future, self._queue.qsize(), self._unfinished_items))
```
[the remark this answers is not in the document]
Comment (JustAnotherArchivist) [reply wpull/pipeline/pipeline.py]: Wrapped this like it's done in wpull/protocol/ftp/command.py already, although I don't like this style.

File: wpull/url.py [hunk 12]
```diff
@@ -127,6 +130,7 @@ def parse(cls, url, default_scheme='http', encoding='utf-8'):
             return None
 
         url = url.strip()
+        # TODO: \r\n\t should be removed from anywhere in the URL.
```
[the remark this answers is not in the document]
Comment (JustAnotherArchivist) [reply wpull/url.py]: The code currently only removes those characters at the beginning and the end, but not within the URL. The relevant issue is #369.
I probably added that TODO for myself and forgot to keep it out of the commit. I'll just remove it.

File: wpull/urlprioritiser.py (hunk 1 above)
Comment (JustAnotherArchivist) [reply wpull/urlprioritiser.py]: I agree. This currently also *always* `deepcopy`s the info and record objects, even if the hook isn't used at all.

There are probably three solutions to this:

1. Implement a `clone` method into `URLInfo` and `URLRecord` to create copies/clones.
2. Add a `freeze` method to these objects to make them immutable. I couldn't find any code that manipulates these objects after the initial creation. But this seems like a pretty good way to reintroduce the performance issues that apparently occurred with `FrozenDict` before its removal in 56138077.
3. Don't do anything about it in the code at all, just passing the original objects – maybe add a note in the documentation to not modify those from the hooks, then ignore it and let people shoot themselves in the foot if they really want to.

I'm tempted to go with number 3.

Comment (ivan) [reply wpull/urlprioritiser.py]: 3 sounds fine with me. I don't think we have enough hooks authors to justify guardrails here.

**Comment by JustAnotherArchivist** [comment]
> s/priorisation/prioritisation/ or the z equivalent in your commit messages as well

Do you want me to edit the commit messages (and thereby rewrite the history of my branch, requiring a force-push) or not?
I guess I'm fine with that, assuming GitHub handles it gracefully. Just want to make sure.

**Comment by JustAnotherArchivist** [comment]
Here we go... For reference, the old branch is preserved at <https://github.com/JustAnotherArchivist/wpull/tree/pr393-old>.

**Comment by JustAnotherArchivist** [comment]
I need to rebase again. Forgot to update the commit IDs in the commit messages.
I also still need to look into https://github.com/ArchiveTeam/wpull/pull/393#issuecomment-435061615.
So not ready for merging yet.

**Comment by JustAnotherArchivist** [comment]
@ivan, could you re-review please?

**Comment by anarcat** [comment]
tests are still red though... :/

**Comment by JustAnotherArchivist** [comment]
@anarcat Yeah, those are due to @falconkirtaran's changes between 2.0.1 and 2.0.3. As mentioned earlier, I only fixed the test failures related to my changes. The remaining ones can be addressed afterwards.

**Comment by anarcat** [comment]
okay, then maybe that can be banged out in #402

**Comment by ivan** [comment]
I was waiting on confirmation that the new queries (which now include priority) didn't regress performance (discussed on IRC;  becoming slower here would be bad) before reviewing this again, but if someone else wants to review this, that is fine with me :-)

**Comment by JustAnotherArchivist** [comment]
With the added index and INDEXED BY clause in d8fee1a8, performance is somewhat worse than before. There really is no way around a small regression since the index is larger than before, so SQLite has more data to crunch on each check-out.

I tested the performance with a large (~10 GiB) DB that I had from a crashed ArchiveBot job. It has over 50 million URLs with roughly 15 million processed. I did not do extensive tests with other databases, but I think this one should cover a fairly extreme case. The test consists of checking out and back in 1000 entries from the DB. The script can be found [here](https://gist.github.com/JustAnotherArchivist/3603136f0547638b826c9edaed999ff4). The same script works with both 2.0.3 and d8fee1a8; it uses code inspection to determine the number of arguments to `check_out` to distinguish between the two.

Averaged across 10 runs each, interleaved, on the same machine: 9.3 seconds for the 2.0.3 code, 12.1 seconds for d8fee1a8. In other words, a performance penalty of about 30 %. That's not too surprising given the increase in index size: an index on just the `status` column covers 11 bytes per row while the new index covers the same 11 bytes for `status` plus 1 byte for `priority` (because it's all zero in this test, another extreme case since the index on `priority` is effectively useless) plus 3 or 4 bytes for `id` (not sure exactly how SQLite switches between the integer representations; cf. [here](https://www.sqlite.org/datatype3.html)) = 15-16 bytes in total or about 45 % more...

This is certainly not a negligible performance impact, but I'm not sure there's any way to reduce it. At the same time, I *am* sure that there are other (and simpler?) ways to improve the database's performance by at least a factor 20 (#427), so I don't think these 30 % matter too much.

I intend to do some more tests with different priority values in the table (instead of everything at zero). I'd expect the performance hit to be less severe then since the new index would be more efficient.

**Comment by JustAnotherArchivist** [comment]
SQLAlchemy 1.3.0 broke my implementation of the `INDEXED BY` clause via <https://github.com/sqlalchemy/sqlalchemy/issues/4509>. So that part will require some more work. Although in the context of #427, getting rid of SQLAlchemy entirely is an option, so I might leave that for later and just limit the SQLAlchemy version to `<1.3.0` if nobody objects.
